### PR TITLE
Return axes instead of plotting for LRFinder

### PR DESCRIPTION
**Pull request by Moh-Yakoub**
Description:

Based on the discussion we had here https://github.com/pytorch/ignite/pull/1930#pullrequestreview-637461553 I propose replacing the plot method with returning axis as seaborn API

Please not I specified the return type of the method as ANY since we don't import the matplotlib library by default underwhich axes are defined

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by vfdev-5** [comment]
@Moh-Yakoub thanks for the PR ! Looks like it is backward compatibility (BC) breaking. We have to think about if it is possible to keep it compatible.

File: ignite/contrib/handlers/lr_finder.py [hunk 1]
```diff
@@ -210,8 +203,9 @@ def plot(
             log_lr: True to plot the learning rate in a logarithmic
                 scale; otherwise, plotted in a linear scale. Default: True.
             filepath: The file name to save the plot to. Default: None.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:205]: filepath was removed from args.

**Review by vfdev-5 — COMMENTED** [review]
Inspecting more the notebook, I think we can still keep it as 
```python
lr_finder.plot()
```
without explicit matplotlib import and `.plot(ax)` call.
The point was talking about seaborn etc is that it can accept `ax` as argument => we can plot Loss curve on a given axis and thus user can embed into other subplots etc.

We are now removing filepath (which is OK), but maybe we could provide an example on how to save the plot on the disk as it was asked in the other issue...

**Comment by Moh-Yakoub** [comment]
> Inspecting more the notebook, I think we can still keep it as
> 
> ```python
> lr_finder.plot()
> ```
> 
> without explicit matplotlib import and `.plot(ax)` call.
> The point was talking about seaborn etc is that it can accept `ax` as argument => we can plot Loss curve on a given axis and thus user can embed into other subplots etc.
> 

@vfdev-5 what do you think of passing an `ax` argument that can be None, and inevitably the user can either use it, or if it's not used, we just plot as the existing behavior. what do you think?

**Comment by vfdev-5** [comment]
> > Inspecting more the notebook, I think we can still keep it as
> > ```python
> > lr_finder.plot()
> > ```
> > 
> > 
> > without explicit matplotlib import and `.plot(ax)` call.
> > The point was talking about seaborn etc is that it can accept `ax` as argument => we can plot Loss curve on a given axis and thus user can embed into other subplots etc.
> 
> @vfdev-5 what do you think of passing an `ax` argument that can be None, and inevitably the user can either use it, or if it's not used, we just plot as the existing behavior. what do you think?

Yes, `ax` arg with `ax=None` by default is something I was thinking about as you say here.
As a return type, we can return `ax` either created internally or the one passed as input arg...

File: ignite/contrib/handlers/lr_finder.py [hunk 2]
```diff
@@ -193,7 +193,7 @@ def plot(
         skip_end: int = 5,
         log_lr: bool = True,
         display_suggestion: bool = True,
-        filepath: Optional[str] = None,
+        axes: Any = None,
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:196]: Let's call it `ax` as in seaborn: https://seaborn.pydata.org/generated/seaborn.lineplot.html#seaborn.lineplot

File: ignite/contrib/handlers/lr_finder.py [hunk 3]
```diff
@@ -212,7 +212,7 @@ def plot(
             log_lr: True to plot the learning rate in a logarithmic
                 scale; otherwise, plotted in a linear scale. Default: True.
             display_suggestion: if True, red dot shows the suggested learning rate.
-            filepath: The file name to save the plot to. Default: None.
+            axes: An axes object used for plotting. Default: None.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:215]:
```suggestion
            ax: Pre-existing axes for the plot. Default: None.
```

File: ignite/contrib/handlers/lr_finder.py [hunk 4]
```diff
@@ -264,18 +263,17 @@ def plot(
             losses = losses[skip_start:-skip_end]
 
         # Plot loss as a function of the learning rate
-        plt.plot(lrs, losses)
+        if axes is None:
+            f, axes = plt.subplots(**kwargs)
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:267]: This part of code should be executed before `if display_suggestion:` and we have to remove `fig, ax = plt.subplots()`, line 252.

File: ignite/contrib/handlers/lr_finder.py [hunk 5]
```diff
@@ -212,7 +212,7 @@ def plot(
             log_lr: True to plot the learning rate in a logarithmic
                 scale; otherwise, plotted in a linear scale. Default: True.
             display_suggestion: if True, red dot shows the suggested learning rate.
-            filepath: The file name to save the plot to. Default: None.
+            axes: An axes object used for plotting. Default: None.
             kwargs: optional kwargs passed to ``plt.savefig`` if ``filepath`` is provided.
                 matplotlib method `savefig` to save the plot.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:217]: I think I already asked you, please add an example code snippet to this docstring on how to save the figure.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update @Moh-Yakoub , some more updates are required and we can merge it

**Comment by vfdev-5** [comment]
As there are more LR finder improvements coming (e.g. #1981), @Moh-Yakoub let's merge this PR before.

cc @KickItLikeShika for visibility.

File: ignite/contrib/handlers/lr_finder.py [hunk 6]
```diff
@@ -193,7 +193,7 @@ def plot(
         skip_end: int = 5,
         log_lr: bool = True,
         display_suggestion: bool = True,
-        filepath: Optional[str] = None,
+        ax: Any = None,
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:196]:
```suggestion
        ax: Optional[Any] = None,
```

File: ignite/contrib/handlers/lr_finder.py [hunk 7]
```diff
@@ -212,9 +212,14 @@ def plot(
...
                 scale; otherwise, plotted in a linear scale. Default: True.
             display_suggestion: if True, red dot shows the suggested learning rate.
-            filepath: The file name to save the plot to. Default: None.
-            kwargs: optional kwargs passed to ``plt.savefig`` if ``filepath`` is provided.
-                matplotlib method `savefig` to save the plot.
+            ax: Pre-existing axes for the plot. Default: None.
+            kwargs: optional kwargs passed to ``plt.subplots`` if ``ax`` is not provided.
+
+        .. code-block:: python
+
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:219]:
```suggestion

```

File: ignite/contrib/handlers/lr_finder.py [hunk 8]
```diff
@@ -212,9 +212,13 @@ def plot(
...
                 scale; otherwise, plotted in a linear scale. Default: True.
             display_suggestion: if True, red dot shows the suggested learning rate.
-            filepath: The file name to save the plot to. Default: None.
-            kwargs: optional kwargs passed to ``plt.savefig`` if ``filepath`` is provided.
-                matplotlib method `savefig` to save the plot.
+            ax: Pre-existing axes for the plot. Default: None.
+            kwargs: optional kwargs passed to ``plt.subplots`` if ``ax`` is not provided.
+
+        .. code-block:: python
+            ax = lr_finder.plot(skip_end=0)
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:219]:
```suggestion
        .. code-block:: python

            ax = lr_finder.plot(skip_end=0)
```

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @Moh-Yakoub

**Pull request merged**